doc/design/xgboost_feature_column.md

```diff
@@ -42,38 +42,75 @@ The training process of XGBoost model inside SQLFlow are as follows:
+SVM file. This method is suitable for offline training (both standalone and distributed), prediction and evaluation, since same
+transformation Python codes can be generated in both training, prediction and evaluation. But it is not suitable for online
+serving, because online serving usually uses other libraries or languages (like C++/Java), which does not support the
+transformation codes we generate in SQLFlow Python codes.
+
+- Method 2. Modify the training iteration of XGBoost, and insert transformation codes during each iteration. But it is not
+easy to modify the training iteration of XGBoost. Moreover, it is also not suitable for online serving for the same reason 
+as method 1.
+
+- Method 3. Combine data pre-processing and model training as a sklearn pipeline. Since sklearn pipeline can be saved as 
+a PMML file by [sklearn2pmml](https://github.com/jpmml/sklearn2pmml) or [Nyoka](https://github.com/nyoka-pmml/nyoka), this
+method is suitable for standalone training, offline prediction, offline evaluation and online serving. Distributed training 
+of sklearn pipeline can be performed using [Dask](https://github.com/dask/dask). However, distributed training pipeline
+cannot be saved as a PMML file directly. It is because Dask does not use but mocks native sklearn APIs to build a pipeline,
+and these mocked APIs cannot be saved. Another problem is that sklearn pipeline only supports very few data pre-processing 
+transformers. For example, hashing a string to a single integer is not supported in sklearn. Of course, we can add more data
+pre-processing transformers, but these new added transformers cannot be saved as a PMML file.
+
+In summary, one of the most critical things is how data pre-processing transformers can be saved for online serving. 
+After investigating the online serving platform in the company (Arks, etc), data pre-processing steps are usually not saved in 
+PMML or Treelite files. The online serving platform provides plugins for users to choose their data pre-processing steps instead
+of loading them from PMML or Treelite files. Therefore, we prefer to choose Method 1 to implement feature column in XGBoost models.
+
+The feature column transformers in Python can be implemented as:
 
 ```Python
 class BaseFeatureColumnTransformer(object):
     def __call__(self, inputs):
         raise NotImplementedError()
+        
+    def set_column_names(self, column_names):
+        self.column_names = column_names
 
 class NumericColumnTransformer(BaseFeatureColumnTransformer):
-    # `column_idx` is the feature index inside `SELECT` statement
-    def __init__(self, column_idx):
-        self.column_idx = column_idx
+    # `key` is the column name inside `SELECT` statement
+    def __init__(self, key):
+        self.key = key
+        
+    def set_column_names(self, column_names):
+        BaseFeatureColumnTransformer.set_column_names(column_names)
+        self.index = self.column_names.index(self.key)
```

Comment:
Why move the `self.index` inside but not pass by an initial argument?

Reply:
Because `self.index` should be inferred from `SELECT` statement. If we want to pass `self.index` inside initial argument, `self.index` must be inferred from `SELECT` statement in Go side, which may be very different from the codegen of TensorFlow and make Go codes more complicated. So in this design, `set_column_names` would be called in Python side when table schema can be obtained in runtime. Maybe it would be more easier.